src/deluge/memory/cache_manager.cpp

```diff
@@ -0,0 +1,176 @@
+
+				// If that previous look was in a different queue, it won't have been included in longestRunSeenInThisQueue, so we have to invalidate that.
+				// TODO: could we just lower it to the longest-run record for that other queue? Yes, done.
+				if (lastTraversalQueue < q && longestRunSeenInThisQueue < longest_runs_[lastTraversalQueue]) {
+					longestRunSeenInThisQueue = longest_runs_[lastTraversalQueue];
+				}
+				stealable = static_cast<Stealable*>(reclamation_queue_[q].getNext(stealable));
+				continue;
+			}
+
+			// If we're forbidden from stealing from a particular thing (usually SampleCache), then make sure we don't
+			if (!stealable->mayBeStolen(thingNotToStealFrom)) {
+				numRefusedTheft++;
+
+				// If we've done this loads of times, it'll be seriously hurting CPU usage. There's a particular case to be careful of - if project
+				// contains just one long pitch-adjusted sound / AudioClip and nothing else, it'll cache it, but after some number of minutes,
+				// it'll run out of new Clusters to write the cache to, and it'll start trying to steal from the cache-Cluster queue, and hit all of these ones
+				// of its own at the same time.
+				if (numRefusedTheft >= 512) {
+					AudioEngine::bypassCulling = true;
+				}
+				stealable = static_cast<Stealable*>(reclamation_queue_[q].getNext(stealable));
+				continue;
+			}
+
+			// If we're not in the last queue, and we haven't tried this too many times yet, check whether it was actually in the right queue
+			if (q < NUM_STEALABLE_QUEUES - 1 && numberReassessed < 4) {
+				numberReassessed++;
+
+				int appropriateQueue = stealable->getAppropriateQueue();
+
+				// If it was in the wrong queue, put it in the right queue and start again with the next one in our queue
+				if (appropriateQueue > q) {
+
+					Debug::print("changing queue from ");
+					Debug::print(q);
+					Debug::print(" to ");
+					Debug::println(appropriateQueue);
+
+					auto* next = static_cast<Stealable*>(reclamation_queue_[q].getNext(stealable));
+
+					stealable->remove();
+					this->QueueForReclamation(appropriateQueue, stealable);
+
+					stealable = next;
+					continue;
+				}
+			}
+
+			// Ok, we've got one Stealable
+			uint32_t* __restrict__ header = (uint32_t*)((uint32_t)stealable - 4);
+			spaceSize = (*header & SPACE_SIZE_MASK);
+
+			stealable->lastTraversalNo = currentTraversalNo;
+
+			// How much additional space would we need on top of this Stealable?
+			int amountToExtend = totalSizeNeeded - spaceSize;
+
+			newSpaceAddress = (uint32_t)stealable;
+
+			// If that one Stealable alone was big enough, that's great
+			if (amountToExtend <= 0) {
+				found = true;
+				break;
+			}
+
+			// Otherwise, see if available neighbouring memory adds up to make enough in total
+			NeighbouringMemoryGrabAttemptResult result = region.attemptToGrabNeighbouringMemory(
+			    stealable, spaceSize, amountToExtend, amountToExtend, thingNotToStealFrom, currentTraversalNo, true);
```

Comment:
I might be missing it but what is thing not to steal from?

Reply:
It's a way to pass around a pointer to the current object or an object that should not be stolen from at any cost when requesting more memory.